nixos/modules/services/torrent/flood.nix

```
@@ -0,0 +1,287 @@
+              type = lib.types.nullOr lib.types.str;
+              default = null;
+              description = lib.mdDoc ''
+                Username of qBittorrent Web API.
+              '';
+            };
+            pass = lib.mkOption {
+              type = lib.types.nullOr lib.types.str;
+              default = null;
+              description = lib.mdDoc ''
+                Password of qBittorrent Web API.
+              '';
+            };
+          };
+          transmission = {
+            url = lib.mkOption {
+              type = lib.types.nullOr lib.types.str;
+              default = null;
+              description = lib.mdDoc ''
+                URL to Transmission RPC interface.
+              '';
+            };
+            user = lib.mkOption {
+              type = lib.types.nullOr lib.types.str;
+              default = null;
+              description = lib.mdDoc ''
+                Username of Transmission RPC interface.
+              '';
+            };
+            pass = lib.mkOption {
+              type = lib.types.nullOr lib.types.str;
+              default = null;
+              description = lib.mdDoc ''
+                Password of Transmission RPC interface.
+              '';
+            };
+          };
+        };
+      };
+    };
+  };
+
+  config =
+    let
+      addAuthOpt = prefix: opt:
+        lib.mapAttrsToList
+          (k: v:
+            "--${prefix}${toString k} ${toString v}")
+          (lib.filterAttrs (k: v: !isNull v) opt);
+      delugeAuth = addAuthOpt "de" cfg.auth.deluge;
+      rtorrentAuth = addAuthOpt "rt" cfg.auth.rtorrent;
+      qbittorrentAuth = addAuthOpt "qb" cfg.auth.qbittorrent;
+      transmissionAuth = addAuthOpt "tr" cfg.auth.transmission;
+      authOpts = lib.lists.findSingle (x: x != [ ]) [ ] [ null ] [ delugeAuth rtorrentAuth qbittorrentAuth transmissionAuth ];
+      args = [
+        "--rundir %S/flood"
+        "--baseuri ${cfg.baseUrl}"
+        "--host ${cfg.address}"
+        "--port ${toString cfg.port}"
+      ]
+      ++ lib.optionals cfg.ssl.enable [ "--ssl" ]
+      ++ lib.optionals (cfg.ssl.enable && (!isNull cfg.ssl.key)) [ "--sslkey ${cfg.ssl.key}" ]
+      ++ lib.optionals (cfg.ssl.enable && (!isNull cfg.ssl.cert)) [ "--sslcert ${cfg.ssl.cert}" ]
+      ++ lib.optionals (authOpts != [ ]) [ "--auth none" ]
+      ++ authOpts
+      ++ map (x: "--allowedpath ${toString x}") cfg.allowedPaths;
+    in
+    lib.mkIf cfg.enable {
+      assertions = [
+        {
+          assertion = authOpts != [ null ];
+          message = "Only one client authentication must be configured";
+        }
+      ];
+      systemd.services.flood =
+        {
+          after = [ "network.target" ];
+          description = "Flood Daemon";
+          wantedBy = [ "multi-user.target" ];
+          path = [ pkgs.mediainfo ];
+          serviceConfig = {
+            ExecStart = "${cfg.package}/bin/flood ${lib.concatStringsSep " " args}";
```

The technically correct one:

```suggestion
            ExecStart = "${cfg.package}/bin/flood ${lib.escapeShellArgs " " args}";
```

---

Actually, I now know that the correct one is `utils.escapeSystemdExecArgs`.